### PR TITLE
Jshook/structure

### PR DESCRIPTION
**Pull request by jshook**
Tangible examples of component nesting and the first demonstrations of practical component lifecycle patterns.

### CODE REVIEW

File: nb-api/src/main/java/io/nosqlbench/adapters/api/util/TagFilter.java [hunk 1]
```diff
@@ -107,6 +105,12 @@ public <T extends Tagged> List<T> filter(List<T> tagged) {
             .collect(Collectors.toList());
     }
 
+    public <T extends NBLabeledElement> List<T> filterLabeled(Collection<T> labeled) {
+        return labeled.stream()
+            .filter(l -> this.matches(l.getLabels().asMap()).matched())
```
Comment (MarkWolters) [inline nb-api/src/main/java/io/nosqlbench/adapters/api/util/TagFilter.java:110]: It seems a little confusing to have both an instance variable and a method named filter (in addition to the streaming filter method)

File: nb-api/src/main/java/io/nosqlbench/adapters/api/util/TagFilter.java [hunk 2]
```diff
@@ -156,6 +160,11 @@ public TagFilter(String filterSpec) {
         }
     }
 
+//    TODO: Implement Me!
```
Comment (MarkWolters) [inline nb-api/src/main/java/io/nosqlbench/adapters/api/util/TagFilter.java:163]: Does this require anything more than 
this.filter.putAll(filterSpec)?

File: nb-api/src/main/java/io/nosqlbench/components/NBComponentSubScope.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.nosqlbench.components;
+
+public class NBComponentSubScope implements AutoCloseable {
```
Comment (MarkWolters) [inline nb-api/src/main/java/io/nosqlbench/components/NBComponentSubScope.java:19]: Not sure I understand the purpose of this class. I see it adding in AutoCloseable but why not simply have that implemented by NBBaseComponent?

**Review by MarkWolters — APPROVED** [review]
Looks good. Didn't leave too many comments as we already walked through this in detail for the most part. Left a couple of comments/questions if you want to respond but nothing that's a showstopper if you want to proceed to merge.

File: nb-api/src/main/java/io/nosqlbench/components/NBComponentSubScope.java (hunk 3 above)
Comment (jshook) [reply nb-api/src/main/java/io/nosqlbench/components/NBComponentSubScope.java:21]: We did implement NBBaseComponent as such, but it seems that this causes some pain for developers with verbose warnings from static analysis anywhere an AutoCloseable is used outside of the try with resources. We'll probably move it to a similar form outside of the official AutoCloseable type.

**Comment by jshook** [comment]
superseded by further work

**Pull request closed without merging**